#!/usr/bin/env python

PACKAGE = "os_lidar_filtering"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# Voxel Grid Downsampling Filter parameters
# gen.add("leaf_size", double_t, 0, "leaf size for voxel grid downsampling", 0.01, 0.0, 0.5)
# The leaf size parameter for voxel grid downsampling is commented out. If enabled, it would allow dynamic adjustment of the voxel grid size, which controls the resolution of the downsampled point cloud.

# Range Filter parameters for intensity filtering
gen.add("min_range", double_t, 0, "Minimum range for pass-through filter", 3.0, 0.0, 1000.0)
# Minimum range for the pass-through filter. Points closer than this distance from the origin are considered noise and removed. Default is 3.0, with a range from 0.0 to 1000.0.

gen.add("max_range", double_t, 0, "Maximum range for pass-through filter", 300.0, 0.0, 1000.0)
# Maximum range for the pass-through filter. Points farther than this distance from the origin are considered noise and removed. Default is 300.0, with a range from 0.0 to 1000.0.

# Statistical Outlier Removal Parameters
gen.add("noise_mean_k", int_t, 0, "Mean K for statistical outlier removal", 75, 1, 100)
# The number of nearest neighbors to use for mean distance estimation in the statistical outlier removal filter. Default is 75, with a range from 1 to 100.

gen.add("noise_stddev", double_t, 0, "Stddev for statistical outlier removal", 3.1, 0.0, 10.0)
# The standard deviation multiplier threshold for determining outliers in the statistical outlier removal filter. Default is 3.1, with a range from 0.0 to 10.0.

# Radius Outlier Removal Parameters
gen.add("radius_search", double_t, 0, "Radius for radius outlier removal", 3.5, 0.0, 10.0)
# The radius within which to search for neighboring points in the radius outlier removal filter. Default is 3.5, with a range from 0.0 to 10.0.

gen.add("min_neighbors", int_t, 0, "Min neighbors for radius outlier removal", 10, 0, 100)
# The minimum number of neighbors required within the radius for a point to be considered an inlier in the radius outlier removal filter. Default is 10, with a range from 0 to 100.

# RANSAC ground plane segmentation Parameters
gen.add("max_iterations", int_t, 0, "Max number of iterations for RANSAC plane removal", 1000, 100, 2000)
# The maximum number of iterations for the RANSAC algorithm used in ground plane segmentation. Default is 1000, with a range from 100 to 2000.

gen.add("ground_distance_threshold", double_t, 0, "Distance threshold for ground plane removal", 0.12, 0.0, 1.0)
# The distance threshold for a point to be considered an inlier of the plane in ground plane segmentation. Default is 0.12, with a range from 0.0 to 1.0.

# Euclidean Extraction Clustering Parameters
gen.add("cluster_tolerance", double_t, 0, "Cluster tolerance for Euclidean Cluster Extraction", 0.1, 0.0, 1.0)
# The maximum distance between points in a cluster for Euclidean Cluster Extraction. Default is 0.1, with a range from 0.0 to 1.0.

gen.add("cluster_min", int_t, 0, "Minimum Cluster Size for Euclidean Cluster Extraction", 100, 0, 15000)
# The minimum number of points required to form a cluster in Euclidean Cluster Extraction. Default is 100, with a range from 0 to 15000.

gen.add("cluster_max", int_t, 0, "Maximum Cluster Size for Euclidean Cluster Extraction", 25000, 0, 50000)
# The maximum number of points allowed in a cluster for Euclidean Cluster Extraction. Default is 25000, with a range from 0 to 50000.

exit(gen.generate(PACKAGE, "os_lidar_filtering_node", "Filter"))